src/main/java/blusunrize/immersiveengineering/common/items/BuzzsawItem.java

```diff
@@ -0,0 +1,775 @@
+			if(state.isIn(BlockTags.LOGS))
+				logs++;
+			else
+			{
+				if(state.isIn(BlockTags.LEAVES))
+					leafTop = true;
+				boolean foundLog = false;
+				if(!leafTop)
+				{
+					// Yay, Acacia trees grow diagonally >_>
+					for(int z = -1; z <= 1; z++)
+						for(int x = -1; x <= 1; x++)
+						{
+							state = world.getBlockState(pos.add(x, 0, z));
+							if(state.isIn(BlockTags.LOGS))
+							{
+								pos = pos.add(x, 0, z);
+								foundLog = true;
+								logs++;
+								break;
+							}
+						}
+				}
+				// If there is no diagonal growth, the tree ends
+				if(!foundLog)
+					break;
+			}
+		}
+		return logs >= 3&&leafTop;
+	}
+
+	private boolean canFellTree(ItemStack stack)
+	{
+		ItemStack sawblade = getSawblade(stack);
+		if(sawblade.getItem() instanceof SawbladeItem)
+			return ((SawbladeItem)sawblade.getItem()).canSawbladeFellTree();
+		return false;
+	}
+
+	/**
+	 * The max distance a block can be from the initial hit
+	 * to still be considered part of the tree
+	 * This is based on the largest vanilla Jungle Trees
+	 */
+	private static final int MAX_HORIZONTAL_DISTANCE = 7;
+
+	private boolean fellTree(World world, BlockPos initialPos, ServerPlayerEntity player, ItemStack stack)
+	{
+		int logs = 0;
+		List<BlockPos> openList = new ArrayList<>();
+		List<BlockPos> closedList = new ArrayList<>();
+		openList.add(initialPos);
+		while(!openList.isEmpty()&&closedList.size() < 512&&logs < 256)
+		{
+			BlockPos next = openList.remove(0);
+
+			// Ignore blocks too far away
+			if(Math.abs(next.getX()-initialPos.getX()) > MAX_HORIZONTAL_DISTANCE
+					||Math.abs(next.getZ()-initialPos.getZ()) > MAX_HORIZONTAL_DISTANCE)
+				continue;
+
+			if(!closedList.contains(next))
+			{
+				BlockState state = world.getBlockState(next);
+				if(state.isIn(BlockTags.LOGS))
+				{
+					closedList.add(next);
+					logs++;
+					// Find all at same level or above, including diagonals
+					for(int y = 0; y <= 1; y++)
+						for(int z = -1; z <= 1; z++)
+							for(int x = -1; x <= 1; x++)
+								openList.add(next.add(x, y, z));
+				}
+				else if(state.isIn(BlockTags.LEAVES))
+				{
+					closedList.add(next);
+					int trunkDist = state.getBlock() instanceof LeavesBlock?state.get(LeavesBlock.DISTANCE): 0;
+					// Leaves only propagate in cardinal directions, and only to other leaves
+					for(Direction dir : new Direction[]{Direction.UP, Direction.NORTH, Direction.SOUTH, Direction.WEST, Direction.EAST})
+					{
+						BlockPos adj = next.offset(dir);
+						BlockState adjState = world.getBlockState(adj);
+						if(adjState.isIn(BlockTags.LEAVES))
+						{
+							int adjDist = adjState.getBlock() instanceof LeavesBlock?adjState.get(LeavesBlock.DISTANCE): 0;
+							if(adjDist < trunkDist) // We don't want to get closer
+								continue;
+						}
+						openList.add(adj);
+					}
+				}
+			}
+		}
+
+		if(closedList.size()==0)
+			return false;
+		// Register a Tick Handler to break the blocks, 10 at a time
```

Comment:
5 or 10? My main reason for disliking this sort of comment, they quickly go out of sync with the code.

Actually, probably 10 because there are events for tick start and tick end, and then more for each loaded dimension.

Reply:
No, you're right, I didn't update the comment. Adding a phase check for tick start.